src/Chainweb/BlockHeader.hs

```diff
@@ -318,19 +318,29 @@ encodeEpochStartTime (EpochStartTime t) = encodeTime t
 decodeEpochStartTime :: MonadGet m => m EpochStartTime
 decodeEpochStartTime = EpochStartTime <$> decodeTime
 
-isEpochStartHeight :: HasChainwebVersion v => v -> BlockHeight -> Bool
-isEpochStartHeight v h = case window (_chainwebVersion v) of
+-- | During the first epoch after genesis there are 10 extra difficulty
+-- adjustments. This is to account for rapidly changing total hash power in this
+-- early stages of the network.
+--
+effectiveWindow :: BlockHeader -> Maybe WindowWidth
+effectiveWindow h = WindowWidth <$> case window ver of
+    Nothing -> Nothing
+    Just (WindowWidth w)
+        | int (_blockHeight h) <= w -> Just $ max 1 $ w `div` 10
+        | otherwise -> Just w
+  where
+    ver = _blockChainwebVersion h
+{-# INLINE effectiveWindow #-}
+
+-- | Return whether the given 'BlockHeader' is the last header in its epoch.
+--
+isLastInEpoch :: BlockHeader -> Bool
+isLastInEpoch h = case effectiveWindow h of
     Nothing -> False
-    Just (WindowWidth n)
-        | h == 0 -> True
-        | int h < n && int h `mod` 10 == (0 :: Int) -> True
-        | int h `mod` n == 0 -> True
+    Just (WindowWidth w)
+        | int (_blockHeight h) `mod` w == 0 -> True
```

Comment:
The last `BlockHeader` in a particular epoch is the one *before* the "mod point". So for an effective window width of 12:
- 0 to 11 form the first epoch/window
- 12 - 23 form the second epoch

Therefore 11 and 23 should report `True` here, not `12` and `24` which is what would occur with `mod w == 0`.

Reply:
DA is expected to occur at the *first* block of the new epoch, so at height 12 and 24.

Reply:
See comment below.